move cip 21 into last call (#209)
celestiaorg/CIPs: cips/cip-21.md

| description | Introduce a new blob type that can be submitted whereby the signer address is included and verified. |
| author | Callum Waters ([@cmwaters](https://github.com/cmwaters)) |
| discussions-to | <https://forum.celestia.org/t/cip-blobs-with-verified-author> |

> | status | Last Call |

Reviewer: Needs a last call date

```suggestion
| status | Last Call |
| last-call-deadline | 2024-10-07 |
```

## Reference Implementation

The go implementation makes the following modifications to the `Blob` proto:

```diff
message BlobProto {
  bytes namespace_id = 1;
  bytes data = 2;
  uint32 share_version = 3;
  uint32 namespace_version = 4;

> + // Signer is sdk.AccAddress that paid for this blob. This field is optional

Reviewer 2: ```suggestion
+ // Signer is sdk.AccAddress that signed this blob. This field is optional
```

Reviewer 2: > So even though the fee granter paid for the blob to be included on-chain, the MsgPayForBlobs submitter (celestia1rgj3z5rfydj3xulhm6glqr0zs7j8rr9rfh6ekw) is the authored blob "signer".

So in this case shouldn't it be `// Signer is sdk.AccAddress that signed this blob`?

+ // and can only be used when share_version is set to 1. 
+ bytes signer = 5;
}
```

The signer is validated in `CheckTx` and `ProcessProposal` as follows:

```go
signer, err := sdk.AccAddressFromBech32(msgPFB.Signer)
if err != nil {
  return err
}

for _, blob := range bTx.Blobs {
  // If share version is 1, assert that the signer in the blob
  // matches the signer in the msgPFB.
  if blob.ShareVersion() == share.ShareVersionOne {
    if appVersion < v3.Version {
      return ErrUnsupportedShareVersion.Wrapf("share version %d is not supported in %d. Supported from v3 onwards", blob.ShareVersion(), appVersion)
    }
    if !bytes.Equal(blob.Signer(), signer) {
      return ErrInvalidBlobSigner.Wrapf("blob signer %s does not match msgPFB signer %s", sdk.AccAddress(blob.Signer()).String(), msgPFB.Signer)
    }
  }
}
```

> 

Reviewer: markdownlint is failing b/c too many empty lines

```suggestion
```

## Reference Implementation

The go implementation makes the following modifications to the `Blob` proto:

```diff
message BlobProto {
  bytes namespace_id = 1;
  bytes data = 2;
  uint32 share_version = 3;
  uint32 namespace_version = 4;

> + // Signer is sdk.AccAddress that paid for this blob. This field is optional

Reviewer: [question] can this feature be used with feegrant? If yes, and we have:

1. celestiaA agrees to pay gas for all message for celestiaB 
2. celestiaB submits an authored blob and uses gas from celestiaA. 
3. Whose address goes in the signer field?

If it's celestiaB then should we stop using the term "that paid for this blob" and instead use something like: 

```go
// Signer is sdk.AccAddress that submitted this blob.
```

Reviewer 2: good catch!

Reviewer 2: After reading through this CIP again recently, I'm a little confused how this would handle blobs submitted with feegrant as described in this thread.
